lib/src/json_serializable/type_helper.dart

```diff
@@ -0,0 +1,83 @@
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/dart/element/type.dart';
+
+abstract class TypeHelper {
+  const TypeHelper();
+
+  bool matchesTypeForSerialize(DartType type);
+
+  bool canDeserialize(DartType type);
+
+  String toSerializedValue(DartType searchType, String name);
+
+  String toDeserializedValue(DartType searchType, String expression);
+}
+
+class JsonHelper extends TypeHelper {
+  const JsonHelper();
+
+  @override
+  bool matchesTypeForSerialize(DartType type) {
+    //TODO(kevmoo): This should be checking for toJson method, but
+    //              toJson might be gone during generation, so check for
+    //              the annotation, too!
+    if (type is! InterfaceType) return false;
+
+    var classElement = type.element as ClassElement;
+
+    for (var ctor in classElement.constructors) {
+      if (ctor.name == 'fromJson') {
+        // TODO: validate that there are the right number and type of arguments
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  bool canDeserialize(DartType type) {
+    if (type is! InterfaceType) return false;
+
+    var classElement = type.element as ClassElement;
+
+    for (var ctor in classElement.constructors) {
+      if (ctor.name == 'fromJson') {
+        // TODO: validate that there are the right number and type of arguments
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  @override
+  String toSerializedValue(DartType searchType, String name) => name;
```

Comment:
So is this just relying on the fact that default json encoding will call toJson? That seems like it is ok but I would like to confirm my understanding. Maybe it is worth adding a comment as well?